core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -317,6 +322,34 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
       }
     }
   }
+  
+  def handleVerifyPartitionsInTransaction(transactionalId: String,
+                                          producerId: Long,
+                                          producerEpoch: Short,
+                                          partitions: collection.Set[TopicPartition],
+                                          responseCallback: VerifyPartitionsCallback): Unit = {
+    if (transactionalId == null || transactionalId.isEmpty) {
+      debug(s"Returning ${Errors.INVALID_REQUEST} error code to client for $transactionalId's AddPartitions request")
+      responseCallback(AddPartitionsToTxnResponse.resultForTransaction(transactionalId, partitions.map(_ -> Errors.INVALID_REQUEST).toMap.asJava))
+    } else {
+      val result: ApiResult[(Int, TransactionMetadata)] = getTransactionMetadata(transactionalId, producerId, producerEpoch, partitions)
+      
+      result match {
+        case Left(err) =>
+          debug(s"Returning $err error code to client for $transactionalId's AddPartitions request")
+          responseCallback(AddPartitionsToTxnResponse.resultForTransaction(transactionalId, partitions.map(_ -> err).toMap.asJava))
+
+        case Right((_, txnMetadata)) =>
+          val txnMetadataPartitions = txnMetadata.topicPartitions
+          val addedPartitions = partitions.intersect(txnMetadataPartitions)
+          val nonAddedPartitions = partitions.diff(txnMetadataPartitions)
+          val errors = mutable.Map[TopicPartition, Errors]()
+          addedPartitions.foreach(errors.put(_, Errors.NONE))
+          nonAddedPartitions.foreach(errors.put(_, Errors.INVALID_TXN_STATE))
```

Comment:
nit: I am not sure if it makes a real difference but did you consider doing something like this:

```
partitions.foreach { tp =>
  if (txnMetadata.topicPartitions.contains(tp))
    ...
  else
   ...
}
```

If works, it would avoid allocating the intermediate collections. I leave this up to you.

Reply:
sure